### DISCUSSION TITLE
Dynamically define a controller

### DISCUSSION THREAD

**Discussion by please-fill-out-this-field** [mojolicious/mojo#2183] [Q&A]
I want to write a Mojolicious plugin that adds a `Foo` controller to the app that doesn't have such controller (i.e. there is no `lib/MyApp/Controller/Foo.pm`).

```perl
package Mojolicious::Plugin::MyPlugin;
use Mojo::Base 'Mojolicious::Plugin', -signatures;

sub register ($self, $app, $conf) {
  my $foo = create_a_controller_somehow('Foo', {
    create => sub ($c) { $c->render(text => 'create') },
    read   => sub ($c) { $c->render(text => 'read') },
    update => sub ($c) { $c->render(text => 'update') },
    delete => sub ($c) { $c->render(text => 'delete') }
  });

  die 'Unable to add: controller already exists'
    if app_already_has_a_controller_with_that_name($app, $foo);

  add_the_controller_to_the_app_somehow($app, $foo)
  $app->routes->post('/foo/create')->to('foo#create');
  ...
}
```

Is it possible without dirty hacks?

Thanks!

**Comment by Akron** [comment]
I think it's better to explain, what you want to achieve in the end. I would think there is no benefit in creating a class. What is your purpose?

If you just want to route to certain methods on plugin registration, just use callbacks like

```perl
$app->routes->post('/foo/create')->to(cb => sub { shift->render(text => 'create' )})
```

If you want to add the routes for several paths or use "action" like routing, you can add a shortcut on registration, like

```perl
$app->routes->add_shortcut(
    mycrud => sub {
      my $r    = shift;
      $r->post('/')->to(
        cb => sub {  shift->render(text => 'create') }
      );
      $r->get('/')->to(
        cb => sub {  shift->render(text => 'read') }
      );
      ....
    }
);
```

**Reply by please-fill-out-this-field** [reply]
>what you want to achieve in the end.

I want a modular app, but instantiating an entire Mojolicious application for a single controller (i.e. Mojolicious::Plugin::Mount) sounds like overkill. I thought maybe there's a better option...

**Reply by guest20** [reply]
People also enjoyed: [`Mojolicious::Plugin::Mojolyst `](https://metacpan.org/release/SADAMS/Mojolicious-Plugin-Mojolyst-0.01/view/lib/Mojolicious/Plugin/Mojolyst.pm)

**Reply by Akron** [reply]
That sounds like registering a shortcut in a plugin is your way to go.
